src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.Cache.cs

```diff
@@ -54,22 +54,35 @@ public abstract partial class JsonTypeInfo
 
         [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
         [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
-        internal JsonPropertyInfo CreatePropertyUsingReflection(Type propertyType, JsonConverter converter)
+        internal JsonPropertyInfo CreatePropertyUsingReflection(Type propertyType)
         {
-            Debug.Assert(propertyType.IsInSubtypeRelationshipWith(converter.TypeToConvert));
+            JsonPropertyInfo? jsonPropertyInfo;
 
-            if (converter.TypeToConvert == propertyType)
+            if (Options.TryGetTypeInfoCached(propertyType, out JsonTypeInfo? jsonTypeInfo))
             {
-                // For the vast majority of use cases, the converter type matches the property type.
-                // Avoid reflection-based initialization by delegating JsonPropertyInfo<T> construction to the converter itself.
-                return converter.CreateJsonPropertyInfo(declaringTypeInfo: this, Options);
+                // If a JsonTypeInfo has already cached for the property type,
+                // avoid reflection-based initialization by delegating construction
+                // of JsonPropertyInfo<T> construction to the property type metadata.
+                return jsonTypeInfo.CreateJsonPropertyInfo(declaringTypeInfo: this);
```

Comment:
note that you might end up with not consistent `Options` instance here since cache might give you type info from equivalent options but reference might be different. That might be unexpected for users

Reply:
You're right. I'll revert the changes to the test and make that particular factory accept an `options` parameter.